google/cloud/spanner_dbapi/connection.py

```diff
@@ -142,18 +174,34 @@ def get_table_column_schema(self, table_name):
     def close(self):
         """Close this connection.
 
-        The connection will be unusable from this point forward.
+        The connection will be unusable from this point forward. Rollback
+        will be performed on all the pending transactions.
         """
+        for transaction in self.transactions:
+            transaction.rollback()
+
         self.__dbhandle = None
         self.is_closed = True
 
     def commit(self):
         """Commit all the pending transactions."""
-        warnings.warn(AUTOCOMMIT_MODE_WARNING, UserWarning, stacklevel=2)
+        if self.autocommit:
+            warnings.warn(AUTOCOMMIT_MODE_WARNING, UserWarning, stacklevel=2)
```

Comment:
Can't say I completely agree on that. If one is calling `commit()` in `autocommit=true` mode, it's definitely a logical error, but I don't think it'll break data or current operations flow. So, as I see it, we can, you know, forgive it, but warningly wave a finger.

In Java you're throwing an error in such a cases?

Reply:
I would recommend that we follow the convention of other Python DB API implementations. So the most important question is what do the implementations for for example PostgreSQL and MySQL do?

The Java implementation throws an exception if you try to call `commit` or `rollback` when in autocommit mode.

Reply:
It looks like every database solves this question by itself. In [SQLite3](https://github.com/coleifer/pysqlite3/blob/master/src/connection.c) I see they are doing nothing if someone called `commit()` in autocommit mode (not even throwing a warning). [MySQL](https://github.com/mysql/mysql-connector-python/blob/5dba2597d9c52f348ee1a3b0b6628f2c035cdc9e/lib/mysql/connector/connection.py) just sends a `COMMIT` statement without any checks. [Postgres](https://github.com/psycopg/psycopg2/blob/f2852a520a26d88fb248c0563e73af43110a13a8/psycopg/pqpath.c) connector prints a warning `no transaction to commit`.

Thus, plus-minus, in Python we usually don't fail on calling commit in autocommit mode. I suppose it's not the mode users often changing, so the chance that one will make a mistake here is low.

Reply:
If we didn't care about following DBAPI conventions I'd agree this should be an error. But sqlite3 is the oldest and likely most-used DBAPI library, and IIRC used as the reference implementation for DBAPI. Following its lead, I think it's fine for this and `rollback` to warn instead of raise.